Remediate BRR-04M

cliffhall opened the pull request:
* In BosonRouter.sol,
  - Added private mapping approvedGates
    - Maps an address to a bool
  - Added LogGateApprovalChanged event
    - params are gate address and approved status
  - Added onlyApprovedGate modifier
    - reverts if approvedGates does not contain a true mapping for the given address
    - added setGateApproval method
      - params are gate address and approved status
      - reverts if approvedGates has the same mapping for the given address as the param
    - added onlyApprovedGate modifer to external methods that take a gate address argument
      - requestCreateOrderETHETHConditional
      - requestCreateOrderTKNTKNWithPermitConditional
      - requestCreateOrderETHTKNWithPermitConditional
      - requestCreateOrderTKNETHConditional
  - In all external methods that take a gate address
    - removed zero address check of gate address that were in some but not all.
    - Now the zero address, like any other address, will revert with the "INVALID_GATE" reason if not approved.

* In 11_gate.ts,
  - in beforeEach of 'Boson router operations'
    - approve gate on boson router
  - added 'Setting gate approval status' context with tests
    - 'Owner should be able to approve a gate'
    - 'Owner should be able to un-approve a gate'
    - '[NEGATIVE] gate approval should revert if not called by owner'
    - '[NEGATIVE] gate approval should revert if owner sends zero address'

* In 12_conditional_commit.ts,
  - in deployContracts function
    - after contract deployment,
      - called setGateApproval on contractBosonRouter with contractGate's address and true for approved.
   - In all negative tests of gate address,
     - changed expected revert reasons EOA and ZERO_ADDRESS_NOT_ALLOWED to INVALID_GATE, which is now the reason in both cases.
     - changed one such test to check that a call with the address of the previously approved contractGate contract would fail if setGateApproval is subsequently called on it with false for approval status.
     - A few of the tests seemed **not to be doing what the test description said**
       - example test at previous line 2007
         - [NEGATIVE] Should revert if specified gate contract does not exist
         - actually successfully called the router with an invalid gate and apparently did not fail because the only check on the contract before was for zero address and this was sending an EOA
         - previously carried on and did other stuff and tested for a method that doesn't take a gate address to revert for EOA reason.
         - I updated to expect revert on initial call to router (via utility function createOrderConditional)

* In events.ts,
  - added LOG_GATE_APPROVAL_CHANGED constant

* In revertReasons.ts,
  - Added INVALID_GATE constant

* In deploy.ts,
  - in deployContracts function,
    - after boson router and gate are both deployed,
      - approve gate on boson router
  
* This fixes #178

cliffhall committed:
Remediate BRR-04M

* In BosonRouter.sol,
  - Added private mapping approvedGates
    - Maps an address to a bool
  - Added LogGateApprovalChanged event
    - params are gate address and approved status
  - Added onlyApprovedGate modifier
    - reverts if approvedGates does not contain a true mapping for the given address
    - added setGateApproval method
      - params are gate address and approved status
      - reverts if approvedGates has the same mapping for the given address as the param
    - added onlyApprovedGate modifer to external methods that take a gate address argument
      - requestCreateOrderETHETHConditional
      - requestCreateOrderTKNTKNWithPermitConditional
      - requestCreateOrderETHTKNWithPermitConditional
      - requestCreateOrderTKNETHConditional
  - In all external methods that take a gate address
    - removed zero address check of gate address that were in some but not all.
    - Now the zero address, like any other address, will revert with the "INVALID_GATE" reason if not approved.

* In 11_gate.ts,
  - in beforeEach of 'Boson router operations'
    - approve gate on boson router
  - added 'Setting gate approval status' context with tests
    - 'Owner should be able to approve a gate'
    - 'Owner should be able to un-approve a gate'
    - '[NEGATIVE] gate approval should revert if not called by owner'
    - '[NEGATIVE] gate approval should revert if owner sends zero address'

* In 12_conditional_commit.ts,
  - in deployContracts function
    - after contract deployment,
      - called setGateApproval on contractBosonRouter with contractGate's address and true for approved.
   - In all negative tests of gate address,
     - changed expected revert reasons EOA and ZERO_ADDRESS_NOT_ALLOWED to INVALID_GATE, which is now the reason in both cases.
     - changed one such test to check that a call with the address of the previously approved contractGate contract would fail if setGateApproval is subsequently called on it with false for approval status.
     - A few of the tests seemed **not to be doing what the test description said**
       - example test at previous line 2007
         - [NEGATIVE] Should revert if specified gate contract does not exist
         - actually successfully called the router with an invalid gate and apparently did not fail because the only check on the contract before was for zero address and this was sending an EOA
         - previously carried on and did other stuff and tested for a method that doesn't take a gate address to revert for EOA reason.
         - I updated to expect revert on initial call to router (via utility function createOrderConditional)

* In events.ts,
  - added LOG_GATE_APPROVAL_CHANGED constant

* In revertReasons.ts,
  - Added INVALID_GATE constant

cliffhall committed:
Remediate BRR-04M

* In deploy.ts,
  - in deployContracts function,
    - after boson router and gate are both deployed,
      - approve gate on boson router

cliffhall committed:
Merge branch 'main' into remediate-BRR-04M

mischat reviewed (changes requested):
This looks good to me. I think we are probably missing a test to ensure the max test coverage... Otherwise i have commented on mainly linting issues. 

I think this is a clean implementation, i wonder whether we should have a public method whereby we can query to ask if an address is a known gate, I know we can reengineer it from the logs, but that can be at times very tedious...

zajck reviewed (changes requested):
I'd like to see also the tests for:

- calling `setGateApproval` and get `No change to approval state`
- Calling `requestCreateOrderXXXYYYConditional` with `INVALID_GATE`

cliffhall committed:
Remediate BRR-04M

- Responding to requests from reviews

* In BosonRouter.sol,
  - removed spurious whitespace
  - removed boolean comparison in onlyApprovedGate
  - indented modifiers on setGateApproval

* In revertReasons.ts,
  - added NO_CHANGE constant

* In 11_gate.ts
  - in 'Boson router operations' context
    - added test '[NEGATIVE] gate approval should revert if no change is represented'

cliffhall committed:
Merge branch 'main' into remediate-BRR-04M

cliffhall committed:
Remediate BRR-04M

- Requests from Heather

* In deploy.ts,
  - moved call to setGateApproval into setDefaults
  - added console log of the address and approved status from the event

cliffhall committed:
Merge branch 'remediate-BRR-04M' of https://github.com/bosonprotocol/contracts into remediate-BRR-04M

cliffhall committed:
Remediate BRR-04M

- Requests from Heather

* In 11_gate.ts,
  - corrected flagged test names to match assertions

cliffhall committed:
Remediate BRR-04M

- make it pass prettier...

mischat reviewed (approved):
LGTM

hswopeams committed:
Merge branch 'main' into remediate-BRR-04M
